### DISCUSSION TITLE
How do I get mypy to work with both pydantic v1.x and pydantic.v1 from v2.x?

### DISCUSSION THREAD

**Discussion by jwodder** [pydantic/pydantic#5971] [Question]
We have several inter-dependent packages that depend on pydantic in ways that will break under v2, and we've decided to stagger the updates to these packages, including having transitional releases that import from either `pydantic.v1` (on pydantic 2.x) or `pydantic` (on pydantic 1.x).  Our code runs correctly with this strategy, but when we try to type-check it with mypy, things break, and I am unsure how to fix this breakage.

An example (available in full at https://github.com/jwodder/pydantic-v1v2-test): This module:

```python
from __future__ import annotations
from pathlib import Path
from typing import Literal

try:
    import pydantic.v1 as pydantic
except ImportError:
    import pydantic  # type: ignore[no-redef]


class Foobar(pydantic.BaseModel):
    classname: Literal["Foobar"] = pydantic.Field("Foobar", readOnly=True)
    foo: int
    bar: str

    @classmethod
    def load(cls, file: Path) -> Foobar:
        return cls.parse_raw(file.read_text())


if __name__ == "__main__":
    print(Foobar(foo=42, bar="hello"))
```

with this mypy config:

```ini
[mypy]
allow_incomplete_defs = False
allow_untyped_defs = False
ignore_missing_imports = True
no_implicit_optional = True
implicit_reexport = False
local_partial_types = True
pretty = True
show_error_codes = True
show_traceback = True
strict_equality = True
warn_redundant_casts = True
warn_return_any = True
warn_unreachable = True
plugins = pydantic.mypy

[pydantic-mypy]
init_forbid_extra = True
warn_untypes_fields = True
```

and this `tox.ini`:

```ini
[tox]
envlist = run,devrun,typing,devtyping
skip_missing_interpreters = True
isolated_build = True
minversion = 3.3.0

[testenv:run]
commands = python -m foobar

[testenv:devrun]
commands_pre =
    pip install --pre --upgrade --no-warn-conflicts git+https://github.com/pydantic/pydantic
commands = python -m foobar

[testenv:typing]
deps =
    mypy
commands =
    mypy src

[testenv:devtyping]
deps = {[testenv:typing]deps}
commands_pre = {[testenv:devrun]commands_pre}
commands = {[testenv:typing]commands}
```

When tested with Python 3.11.3 on macOS 11.7.2 using mypy 1.3.0, using variously pydantic 1.10.8 and dev pydantic commit [f6479b8](https://github.com/pydantic/pydantic/commit/f6479b87b12c90ef87c7bce5c6e9e9cb368cf9ac), the `run` and `devrun` environments succeed, but the `typing` and `devtyping` do not.  The `typing` environment fails with:

```text
src/foobar.py:24: error: Returning Any from function declared to return
"Foobar"  [no-any-return]
            return cls.parse_raw(file.read_text())
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Found 1 error in 1 file (checked 1 source file)
```

and the `devtyping` environment fails with:

```text
src/foobar.py:28: error: Missing named argument "classname" for "Foobar" 
[call-arg]
        print(Foobar(foo=42, bar="hello"))
              ^~~~~~~~~~~~~~~~~~~~~~~~~~~
Found 1 error in 1 file (checked 1 source file)
```

How can this be fixed?

**Comment by dmontagu** [comment]
This is a good question. Could you please create an issue asking this? I'm thinking _maybe_ there's a way to do it with the mypy plugin but that won't work for vscode. And I think it's a reasonable request. Maybe someone will have a good answer but I think it's worth tracking as an issue, not just a question here.

**Reply by jwodder** [reply]
Issue created: https://github.com/pydantic/pydantic/issues/6022
